### PR TITLE
add choice: upper, lower, None

### PR DESCRIPTION
**Pull request by christophmluscher**
Addressing: #177

### CODE REVIEW

File: corpus/stats.py [hunk 1]
```diff
@@ -32,18 +40,46 @@ def tasks(self):
...
+                    for orth in tree.findall(".//lemma/orth")
+                    if orth.text
+                }
+            elif self.casing == "lower":
+                iv_words = {
+                    orth.text.lower()
+                    for orth in tree.findall(".//lemma/orth")
+                    if orth.text
+                }
+            else:
```
Comment (michelwi) [inline corpus/stats.py:55]: maybe check for `None` (or do you mean `"none"`?) explicitly and add an assertion that the user does not request an unknown casing mode.

**Review by michelwi — COMMENTED** [review]
thats a lot of code duplication.. maybe we can define a function `some_name(w)` that returns `w.upper()`, `w.lower()` or `w` depending on mode and then use this instead of the case distinctions.

**Review by curufinwe — CHANGES_REQUESTED** [review]
Better do something like:
```
modifier = lambda x: x
if casing == "upper":
  modifier = lambda x: x.upper()
elif casing == "lower":
  modifier = lambda x: x.lower()
```
And then use modifier instead of branching during creation of iv_words and oov_words

**Comment by christophmluscher** [comment]
> Better do something like:
> 
> ```
> modifier = lambda x: x
> if casing == "upper":
>   modifier = lambda x: x.upper()
> elif casing == "lower":
>   modifier = lambda x: x.lower()
> ```
> 
> And then use modifier instead of branching during creation of iv_words and oov_words

Yes. I agree.
One question/modification. I thought that assigning a lambda statement is discouraged!? I checked PyCharm also complains about this. So I would use a `def` instead of `lambda`.

**Comment by curufinwe** [comment]
Sure, no problem.

**Review by michelwi — APPROVED** [review]

File: corpus/stats.py [hunk 2]
```diff
@@ -16,24 +16,44 @@ class ExtractOovWordsFromCorpusJob(Job):
     Extracts the out of vocabulary words based on a given corpus and lexicon
     """
 
-    def __init__(self, bliss_corpus, bliss_lexicon):
+    __sis_hash_exclude__ = {
+        "casing": "upper",
+    }
+
+    def __init__(self, bliss_corpus, bliss_lexicon, casing="none"):
```
Comment (curufinwe) [inline corpus/stats.py:23]: Why is default "none" but sis_hash_exclude has "upper"? This will break old setups.

Comment (christophmluscher) [reply corpus/stats.py:23]: yes. I missed that.

When I think about this.. the old behavior was upper. so the default value and sis_hash_exclude should be "upper" correct?
@JackTemaki you agree?

Comment (christophmluscher) [reply corpus/stats.py:23]: but we set this in the Librispeech pipeline to "upper".. and that for that case the hash should not change. but for every new setup we want the "none" behavior.. not sure how to handle this...

Comment (JackTemaki) [reply corpus/stats.py:23]: But would it not be correct to have "upper" as default, and exclude this from the hash? Then it does not change anything anywhere.

Comment (JackTemaki) [reply corpus/stats.py:23]: >  but for every new setup we want the "none" behavior
New behavior newer works without breaking old stuff, this is why RETURNN got this `behavior_version` parameter...

Comment (JackTemaki) [reply corpus/stats.py:23]: What if we add a "second" version of the job with changed defaults...? I have no better idea....

Comment (JackTemaki) [reply corpus/stats.py:23]: It would inherit completely from the first one except for that single change

Comment (christophmluscher) [reply corpus/stats.py:23]: maybe have "upper" as default but add a warning for command line or log file!?

Comment (christophmluscher) [reply corpus/stats.py]: "upper" is now default and included in sis_hash_exclude to not break legacy settings. but I added a warning

Comment (christophmluscher) [reply corpus/stats.py]: maybe move the warning to the run function??

**Review by curufinwe — CHANGES_REQUESTED** [review]
Move warning into run function.

**Review by curufinwe — APPROVED** [review]

**Review by curufinwe — APPROVED** [review]

**Pull request merged**